microceph/ceph/config.go

```diff
@@ -3,58 +3,193 @@ package ceph
 	"strings"
 
+	"github.com/canonical/microceph/microceph/api/types"
 	"github.com/canonical/microceph/microceph/common"
 	"github.com/canonical/microceph/microceph/database"
 )
 
+type ConfigTable map[string]struct{
+	Who     string // Ceph Config internal <who> against each key
+	Daemons []string // Daemons that need to be restarted post config change.
+}
+
+// Check if certain key is present in the map.
+func (c ConfigTable) isKeyPresent(key string) bool {
+	if _, ok := c[key]; !ok {
+		return false
+	}
+	
+	return true
+}
+
+// Return keys of the given set
+func (c ConfigTable) Keys() (keys []string) {
+	for k := range c {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+func GetConfigTable() ConfigTable {
+	return ConfigTable{
+		"public_network":  {"global", []string{"mon", "osd"}},
+		"cluster_network": {"global", []string{"osd"}},
+	}
+}
+
+// Struct to get Config Items from config dump json output.
+type ConfigDumpItem struct{
+	Section string
+	Name    string
+	Value   string
+}
+type ConfigDump []ConfigDumpItem
+
+func SetConfigItem(c types.Config) error {
+	configTable := GetConfigTable()
+	
+	args := []string{
+		"config",
+		"set",
+		configTable[c.Key].Who,
+		c.Key,
+		c.Value,
+		"-f",
+		"json-pretty",
+	}
+	
+	_, err := processExec.RunCommand("ceph", args...)
+	if err != nil {
+		return err
+	}
+	
+	return nil
+}
+
+func GetConfigItem(c types.Config) (types.Configs, error) {
+	var err error
+	configTable := GetConfigTable()
+	ret := make(types.Configs, 1)
+	who := "mon"
+
+	// workaround to query global configs from mon entity
+	if configTable[c.Key].Who != "global" {
+		who = configTable[c.Key].Who
+	}
+	
+	args := []string{
+		"config",
+		"get",
+		who,
+		c.Key,
+	}
+	
+	ret[0].Key = c.Key
+	ret[0].Value, err = processExec.RunCommand("ceph", args...)
+	if err != nil {
+		return nil, err
+	}
+	
+	return ret, nil
+}
+
+func RemoveConfigItem(c types.Config) error {
+	configTable := GetConfigTable()
+	args := []string{
+		"config",
+		"rm",
+		configTable[c.Key].Who,
+		c.Key,
+	}
+	
+	_, err := processExec.RunCommand("ceph", args...)
+	if err != nil {
+		return err
+	}
+	
+	return nil
+}
+
+func ListConfigs() (types.Configs, error) {
+	var dump ConfigDump
+	var configs types.Configs
+	configTable := GetConfigTable()
+	args := []string{
+		"config",
+		"dump",
+		"-f",
+		"json-pretty",
+	}
+	
+	output, err := processExec.RunCommand("ceph", args...)
+	if err != nil {
+		return configs, err
+	}
+	
+	json.Unmarshal([]byte(output), &dump)
+	// Only take configs permitted in config table.
+	for _, configItem := range dump {
+		if configTable.isKeyPresent(configItem.Name) {
+			configs = append(configs, types.Config{
+				Key: configItem.Name,
+				Value:  configItem.Value,
+			})
+		}
+	}
+	
+	return configs, nil
+}
+
+// updates the ceph config file.
 func updateConfig(s common.StateInterface) error {
 	confPath := filepath.Join(os.Getenv("SNAP_DATA"), "conf")
 	runPath := filepath.Join(os.Getenv("SNAP_DATA"), "run")
-
+	
```

Comment:
Nit: please clean up spurious white space changes here and below

Reply:
ack